lib/ui/painting/resource_context.cc

```diff
@@ -5,21 +5,42 @@
 #include "flutter/lib/ui/painting/resource_context.h"
 
 #include "lib/fxl/logging.h"
+#include "lib/fxl/synchronization/mutex.h"
 
 namespace blink {
 namespace {
 
 static GrContext* g_context = nullptr;
+static fxl::Mutex g_mutex;
+static volatile bool g_freeze = false;
 
 }  // namespace
 
+ResourceContext::ResourceContext() { g_mutex.Lock(); }
+
+ResourceContext::~ResourceContext() { g_mutex.Unlock(); }
+
 void ResourceContext::Set(GrContext* context) {
   FXL_DCHECK(!g_context);
   g_context = context;
 }
 
 GrContext* ResourceContext::Get() {
-  return g_context;
+  return g_freeze ? nullptr : g_context;
+}
+
+std::unique_ptr<ResourceContext> ResourceContext::Acquire() {
+  return std::make_unique<ResourceContext>();
+}
+
+void ResourceContext::Freeze() {
+  fxl::MutexLocker lock(&g_mutex);
```

Comment:
I agree that this version works. In this model, the IO thread will continue to run, and any pending image loads will continue, but they'll just become SkImages holding encoded CPU data (JPG, PNG, etc...). The alternative would be to have Freeze just call g_mutex.Lock(), and Unfreeze call g_mutex.Unlock(). That would effectively halt the IO thread while the app is backgrounded, and then parallel decoding would resume once it was foregrounded again. I don't have a strong opinion either way - just wanted to point out the implications.

Reply:
That would be a better solution for the specific bug, but as @chinmaygarde pointed out, we probably don't want to block the "IO thread" as there might be other tasks waiting to be processed.

Trying to summarize the other alternatives we considered:
1.  Have an "Image decoding thread" which we should be ok to block in this case.
2.  Fail the image decoding (pass back a failure to the framework), and then re-try when the app is resumed.

Option 2 adds some complexity in the framework that we prefer to avoid, and option 1 didn't have much fans...

I believe that this scenario is pretty edgy (iOS app was backgrounded right after image decoding started),
and assuming that the jank would not be too bad (do you have some hunch here?) I would think it is reasonable to stick with the simpler solution and trade the crash with some potential jank on app resume.

Would be nice if we had an easy way to limit this only to the iOS devices that actually crash.

Reply:
Thanks for reviewing.

So I think the hypothetical scenario we're worried about here is that we'll chain up a whole bunch of SkImages that don't have decoded textures and then come back to the foreground and things will look bad. 

Perhaps we can stretch this through both space (things that fit in a screen) and time (animated gif) dimensions. But I tested it using the worst supported iOS device and had a super long animated gif that almost fully advanced in the background. Coming back to the foreground seemed ok performance wise. Nothing too noticeable.

Either way, I think a semantic fix (don't continue decoding) may be orthogonal to this. i.e. I think we want the Dart side image decoding triggering to respond semantically to application life cycles, not to underlaying GL commands being refused. We'd want some explicit Dart-side logic that stops time on animated gif from advancing when the user can't watch it. Maybe https://github.com/flutter/flutter/pull/13106 already does it.

Reply:
Sorry, @amirh pointed out that this analysis is flawed. I actually tested the engine in a broken state https://github.com/flutter/flutter/issues/13157. The right behaviour (not continuing to decode following frames) is already the existing pre-bug behaviour.